### PR TITLE
Creating new notebook for training compiler TensorFlow support. Multi…

### PR DESCRIPTION
**Pull request by BruceChang9688**
…-GPU-Single-Node

*Issue #, if available:*
#3456 

*Description of changes:*
Notebook trains a Vision Transformer on the Caltech-256 dataset using SageMaker Training Compiler.

*Testing done:*
Tested on SageMaker Notebook instance with conda_tensorflow2_py36 kernel.
[TF_29_single_node_multi_gpu.pdf](https://github.com/aws/amazon-sagemaker-examples/files/8950912/TF_29_single_node_multi_gpu.pdf)

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `tox -e black-format,black-nb-format`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Lokiiiiii** [comment]
Remove all outputs from the notebook in the PR to make it easy to review. The saved copy you attached has the outputs for perusal.

File: sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb [hunk 1]
```diff
@@ -0,0 +1,1080 @@
+{
```
Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>Single-Node with Multiple GPUs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p><code>conda_tesorflow_p39</code> -&gt; <code>conda_tensorflow_p39</code></p><p> </p><p>instance with multiple GPUs -&gt; instance with 4 GPUs</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>sagemaker&gt;=2.95</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>num_gpus_per_instance -&gt; NUM_GPUS (for consistency)</p><p>EPOCHS = 10 * NUM_GPUS (to run the training long enough to amortize the cost of compilation)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>memory of a <code>ml.p3.8xlarge</code> instance -&gt; memory of the V100 GPUs in a <code>ml.p3.8xlarge</code> instance</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>delete the commented out code.</p><p> </p><p>Move the note about SageMaker metric resolution to the markdown under Analysis</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

File: sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
```
Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>we set NUM_GPUS for 4 -&gt; we set NUM_GPUS to 4</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>Remove ToDo</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>Refer to https://github.com/aws/amazon-sagemaker-examples/blob/main/sagemaker-training-compiler/tensorflow/single_gpu_single_node/vision-transformer.ipynb for the waiting section.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>Add the following:</p><p>Note that SageMaker collects metrics at a resolution of 1 minute and metrics for some epochs might be lost.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (Lokiiiiii) [inline sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>We can see that the model's convergence behavior is similar with and without Training Compiler. -&gt; We can see that the model converges faster with SageMaker Training Compiler.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3464/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

**Comment by Lokiiiiii** [comment]
Use the following code to fix formatting errors

```
cd sagemaker-training-compiler/tensorflow/multiple_gpu_single_node
black -l 100 .
```

File: sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
```
[the remark this answers is not in the document]
Comment (Lokiiiiii) [reply sagemaker-training-compiler/tensorflow/multiple_gpu_single_node/vision-transformer.ipynb:1]: <p>Let's stick with <code>SageMaker Training Compiler</code></p>

**Review by jkroll-aws — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Comment by mchoi8739** [comment]
We now allow and encourage to keep the outputs of code cells. Would you rerun and take a snapshot of it with outputs? I'd like to see how the result section looks.

**Comment by mchoi8739** [comment]
> > We now allow and encourage to keep the outputs of code cells. Would you rerun and take a snapshot of it with outputs? I'd like to see how the result section looks.
> 
> @mchoi8739 The resultant notebook with outputs is attached to the PR in the description.

I just saw the pdf, and the result plots looks great! Much better to see what's going on by seeing the plots that lift all questions about why you chose smaller batch size for compiler and how the result will look like. I just want to ask if you don't mind to update this PR with the output plots kept, so that readers can be convinced by the outcome.

**Comment by jkroll-aws** [comment]
> > > We now allow and encourage to keep the outputs of code cells. Would you rerun and take a snapshot of it with outputs? I'd like to see how the result section looks.
> > 
> > 
> > @mchoi8739 The resultant notebook with outputs is attached to the PR in the description.
> 
> I just saw the pdf, and the result plots look great! Much better to see what's going on by seeing the plots that lift all questions about why you chose smaller batch size for compiler and how the result will look like. I just want to ask if you don't mind to update this PR with the output plots kept, so that readers can be convinced by the outcome.

It would be great if you can add inline PNG images of any result plots rather than using notebook outputs, thanks!

**Comment by mchoi8739** [comment]
We can bring this discussion offline. If it's not true that we allow outputs, then I'm fine with the current version. The SMTC team and I can make a follow-up PR later.
